pkg/gameservers/missing.go

```diff
@@ -0,0 +1,157 @@
+	"agones.dev/agones/pkg/util/logfields"
+	"agones.dev/agones/pkg/util/runtime"
+	"agones.dev/agones/pkg/util/workerqueue"
+	"github.com/heptiolabs/healthcheck"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	typedcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	corelisterv1 "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+)
+
+// MissingPodController watches makes sure that any GameServer
+// that isn't in Scheduled, or Unhealthy/Error state, that is missing a Pod
+// moved to Unhealthy. This can sometimes happen if the controller has downtime
+// for a unexpected reason, or if there is no Delete event for a Pod for any reason.
+// Since resync is every 30 seconds, even if there is some time in which a GameServer
+// is in a broken state, it will eventually move to Unhealthy, and get replaced (if in a Fleet).
+type MissingPodController struct {
+	baseLogger       *logrus.Entry
+	podSynced        cache.InformerSynced
+	podLister        corelisterv1.PodLister
+	gameServerSynced cache.InformerSynced
+	gameServerGetter getterv1.GameServersGetter
+	gameServerLister listerv1.GameServerLister
+	workerqueue      *workerqueue.WorkerQueue
+	recorder         record.EventRecorder
+}
+
+// NewMissingPodController returns a MissingPodController
+func NewMissingPodController(health healthcheck.Handler,
+	kubeClient kubernetes.Interface,
+	agonesClient versioned.Interface,
+	kubeInformerFactory informers.SharedInformerFactory,
+	agonesInformerFactory externalversions.SharedInformerFactory) *MissingPodController {
+
+	podInformer := kubeInformerFactory.Core().V1().Pods().Informer()
+	gameServers := agonesInformerFactory.Agones().V1().GameServers()
+	c := &MissingPodController{
+		podSynced:        podInformer.HasSynced,
+		podLister:        kubeInformerFactory.Core().V1().Pods().Lister(),
+		gameServerSynced: gameServers.Informer().HasSynced,
+		gameServerGetter: agonesClient.AgonesV1(),
+		gameServerLister: gameServers.Lister(),
+	}
+
+	c.baseLogger = runtime.NewLoggerWithType(c)
+	c.workerqueue = workerqueue.NewWorkerQueue(c.syncGameServer, c.baseLogger, logfields.GameServerKey, agones.GroupName+".MissingPodController")
+	health.AddLivenessCheck("gameserver-missing-pod-workerqueue", healthcheck.Check(c.workerqueue.Healthy))
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(c.baseLogger.Debugf)
+	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: kubeClient.CoreV1().Events("")})
+	c.recorder = eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: "missing-pod-controller"})
+
+	gameServers.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(_, newObj interface{}) {
+			gs := newObj.(*agonesv1.GameServer)
+			_, isDev := gs.GetDevAddress()
```

Comment:
nit: can be written as

`if _, isDev := gs.GetDevAddress(); !isDev && !isBeforePodCreated(gs) && !gs.IsBeingDeleted() && !gs.IsUnhealthy() { ... }`

to reduce the scope of the `isDev` variable.

Reply:
I'm coming back for a secondary review. I'm not sure if moving this into the one line makes it harder to read or not.

Current:
```golang
       gameServers.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
		UpdateFunc: func(_, newObj interface{}) {
			gs := newObj.(*agonesv1.GameServer)
			_, isDev := gs.GetDevAddress()
			if !isDev && !isBeforePodCreated(gs) && !gs.IsBeingDeleted() &&
				!(gs.Status.State == agonesv1.GameServerStateUnhealthy) {
				c.workerqueue.Enqueue(gs)
			}
		},
	})
```

becomes this:

```golang
	gameServers.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
		UpdateFunc: func(_, newObj interface{}) {
			gs := newObj.(*agonesv1.GameServer)
			if _, isDev := gs.GetDevAddress();
			!isDev && !isBeforePodCreated(gs) && !gs.IsBeingDeleted() &&
				!(gs.Status.State == agonesv1.GameServerStateUnhealthy) {
				c.workerqueue.Enqueue(gs)
			}
		},
	})
```

WDYT?

Reply:
I find it easier because I'm used to that style.